nixos/vte: use vte without any GUI dependencies; nixos/no-x-libs: add vte

SuperSandro2000 opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

SuperSandro2000 committed:
vte: allow to build without any gtk

SuperSandro2000 committed:
nixos/no-x-libs: add vte

jtojnar commented:
I do not think it makes sense to build VTE without GTK, even though the build system currently allows it. VTE is GTK-based library and without choosing GTK version, nothing but a few ancillary files that are pointless without VTE are installed.

SuperSandro2000 commented:
> a few ancillary files that are pointless without VTE are installed.

If they are alone but not when you need them on a remote machine to which you connect over ssh.

Right now I only need vte for https://github.com/NixOS/nixpkgs/blob/177b80f61d05794615e9b2244b183a5c95239699/nixos/modules/config/vte.nix#L9-L13

jtojnar commented:
Could you maybe mention it in the commit message? And note that it will likely break in the future.

SuperSandro2000 commented:
added

SuperSandro2000 committed:
nixos/vte: use vte without any GUI dependencies

we just need the script, borrowed from environment.noXLibs
On my system this reduced the closure size by ~100MB

 ▶ nix path-info -Sh /nix/var/nix/profiles/system-76-link/
/nix/store/adybwlzyf3qa56irabblzlyjsr74amb9-nixos-system-francium-23.11    6.5G
 ▶ nix path-info -S /nix/var/nix/profiles/system-76-link/
/nix/store/adybwlzyf3qa56irabblzlyjsr74amb9-nixos-system-francium-23.11  6956572064

 ▶ nix path-info -Sh /nix/var/nix/profiles/system-77-link/
/nix/store/wgs6wlq2i911q2r1n5fqbs9vzmp8qy26-nixos-system-francium-23.11    6.4G
 ▶ nix path-info -S /nix/var/nix/profiles/system-77-link/
/nix/store/wgs6wlq2i911q2r1n5fqbs9vzmp8qy26-nixos-system-francium-23.11  6856836056

SuperSandro2000 commented:
@bobby285271 @jtojnar PTAL

jtojnar commented on pkgs/development/libraries/vte/default.nix:
```diff
@@ -73,9 +74,10 @@ stdenv.mkDerivation rec {
 
-  propagatedBuildInputs = assert (gtkVersion == "3" || gtkVersion == "4"); [
+  propagatedBuildInputs = assert (gtkVersion == "3" || gtkVersion == "4" || gtkVersion == null);
     # Required by vte-2.91.pc.
-    (if gtkVersion == "3" then gtk3 else gtk4)
+    lib.optional (gtkVersion != null) (if gtkVersion == "3" then gtk3 else gtk4)
+  ++ [
```
Maybe something like this would be cleaner:

```suggestion
    # Required by vte-2.91.pc.
    propagatedBuildInputs = lib.optionals (gtkVersion != null) [
      (assert (gtkVersion == "3" || gtkVersion == "4";
      if gtkVersion == "3" then gtk3 else gtk4)
    ] ++ [
```

jtojnar commented:
Actually, it no longer appears to build.

SuperSandro2000 commented:
Works for me when building it with ``nix-build -E 'with import ./. {}; vte.override { gtkVersion = null; }'``

jtojnar commented on pkgs/development/libraries/vte/default.nix:
```diff
@@ -75,9 +76,11 @@ stdenv.mkDerivation rec {
+    (assert (gtkVersion == "3" || gtkVersion == "4");
+    if gtkVersion == "3" then gtk3 else gtk4)
+  ] ++ [
     glib
     pango
```
These do not need to be propagated without gtk, which would also make the dev output redundant but I am not sure how to get rid of it without adding complexity of having to repeat it in `buildInputs`.

SuperSandro2000 replied:
I've just added pangop to buildInputs.

jtojnar replied:
We would still want to add glib to not rely on propagation.

SuperSandro2000 commented:
@jtojnar can we now merge this?

SuperSandro2000 commented:
@jtojnar can we merge this?
